eventsourcing/src/main/java/org/axonframework/eventsourcing/MultiStreamableMessageSource.java

```diff
@@ -0,0 +1,384 @@
+
+                while (it.hasNext()) {
+                    Map.Entry<String, BlockingStream<TrackedEventMessage<?>>> current = it.next();
+
+                    if (it.hasNext()) {
+                        if (current.getValue().hasNextAvailable()) {
+                            return true;
+                        }
+                    } else {
+                        if (current.getValue().hasNextAvailable((int) Math
+                                .min(longPollTime, deadline - System.currentTimeMillis()), unit)) {
+                            return true;
+                        }
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        /**
+         * Checks each stream to see if a message is available. If more than one stream has a message it decides which
+         * message to return using the {@link #comparator}.
+         *
+         * @return selected message using {@link #comparator}.
+         *
+         * @throws InterruptedException thrown if polling is interrupted during polling a stream.
+         */
+        @Override
+        public TrackedEventMessage<?> nextAvailable() throws InterruptedException {
+            //Return peekedMessage is available.
+            if (peekedMessage != null) {
+                TrackedEventMessage next = peekedMessage;
+                peekedMessage = null;
+                return next;
+            }
+
+            HashMap<String, TrackedSourcedEventMessage> candidateMessagesToReturn = new HashMap<>();
+
+            //block until there is a message to consume
+            while (candidateMessagesToReturn.size() == 0) {
+                peekForMessages(candidateMessagesToReturn);
+            }
+
+            //get streamId of message to return after running the comparator
+            String streamIdOfMessage =
+                    candidateMessagesToReturn.entrySet().stream().min((m1, m2) -> comparator
+                            .compare(m1.getValue(), m2.getValue()))
+                                             .map(Map.Entry::getKey).orElse(null);
+
+            // Actually consume the message from the stream
+            TrackedEventMessage<?> messageToReturn = messageSources.get(streamIdOfMessage).nextAvailable();
+
+            //update the composite token
+            final MultiSourceTrackingToken newTrackingToken = trackingToken.advancedTo(streamIdOfMessage,
+                                                                                       messageToReturn.trackingToken());
+            trackingToken = newTrackingToken;
+
+            logger.debug("Message consumed from stream: {}", streamIdOfMessage);
+            return new GenericTrackedEventMessage<>(newTrackingToken, messageToReturn);
+        }
+
+        private void peekForMessages(HashMap<String, TrackedSourcedEventMessage> candidateMessagesToReturn) {
```

Comment:
I'd change the parameter to `Map<String, TrackedSourcedEventMessage>` i.o. `HashMap<String, TrackedSourcedEventMessage>`.